core/src/main/java/com/graphhopper/routing/querygraph/QueryGraph.java

```diff
@@ -592,110 +309,61 @@ public EdgeExplorer createEdgeExplorer(final EdgeFilter edgeFilter) {
         }
     }
 
-    private EdgeExplorer createUncachedEdgeExplorer(EdgeFilter edgeFilter) {
-        // Iteration over virtual nodes needs to be thread safe if done from different explorer
-        // so we need to create the mapping on EVERY call!
-        // This needs to be a HashMap (and cannot be an array) as we also need to tweak edges for some mainNodes!
-        // The more query points we have the more inefficient this map could be. Hmmh.
```

Comment:
This comment is the reason why I did not merge this PR yet. I dont understand it :) First, we are creating `QueryGraph` per request and we do not use concurrent/multi-threaded edge explorers on the same `QueryGraph`. Second, I did not understand why we built a map of `VirtualEdgeIterator` (with the mutable `current` indices). Why not just a map of `VirtualEdgeIteratorState` ? (ok also not immutable because of `unfavored` flags, but almost..).

Reply:
Good questions :) I think the "thread-safety" is regarding all downstream read-operations (not used here). In this case it was the multi-threaded matrix algorithm but could be also multi-threaded bidirectional dijkstra that I wanted to support.

Regarding the VirtualEdgeIteratorState: how would you implement the custom EdgeExplorer with this?

Reply:
> I think the "thread-safety" is regarding all downstream read-operations (not used here).

Ok that explains it for me / makes sense

> Regarding the VirtualEdgeIteratorState: how would you implement the custom EdgeExplorer with this?

I think the main thread-safety issue in master is that we build a map of `VirtualEdgeIterator` which is just a `List<VirtualEdgeIteratorState>` and an index (`current`). So there is lots of mutable state because of the `current` indices: When one thread calls `next()` on some iterator it will affect the behavior of the same iterator in another thread. This is problem can be easily solved: Just use a map of `List<VirtualEdgeIteratorState>` (without the `current` indices) and keep the `current` indices within the scope of the single threads (one `VirtualEdgeIterator` per thread, but they all share the `List<VirtualEdgeIteratorState>`s).  You can see this in this PR, I have already done this.
If these lists were immutable the code would be thread-safe. Unfortunately the  `VirtualEdgeIteratorState`s are not immutable, so we would either have to work towards making them immutable or use them in a thread-safe way (do not change any of the mutable state).

Reply:
Btw I think the query graph cache is also not thread-safe ?

Reply:
> Btw I think the query graph cache is also not thread-safe ?

Oh, yes. 

> You can see this in this PR, I have already done this.

Ok, so this means QueryGraph (incl. the cache) is finally read-thread safe :) (?)

IMO everything in GraphHopper should be read-thread safe and write thread-safe is currently forced via locking: https://github.com/graphhopper/graphhopper/blob/master/core/src/main/java/com/graphhopper/GraphHopper.java#L965

Reply:
> Ok, so this means QueryGraph (incl. the cache) is finally read-thread safe :) (?)

Hmm in which sense do you think it was not read-thread safe before ?

Reply:
I just misinterpreted the code and thought the iterator.reset was no longer used. So I think it is still not "read-thread-safe" in the sense that the iterator safes state (getCurrentEdge)

Reply:
Yes the iterator safes state (it has to), but the `currentEdge` is no longer part of the map. Its thread-safe in the sense that every thread gets its own iterator instance (and hence its own `currentEdge` index).